### PR TITLE
Update gcloud module to support GCS Buckets

### PR DESCRIPTION
**Pull request by CarlosRS9**
Hello team,

This PR is related to #8716. It will add support for GCP logs coming from a bucket, as the module only supports PubSub logs at the moment.

## Changes

Several changes needed to be applied to the module structure in order to add support for the buckets while maintaining PubSub compatibility and ensuring that in the future it will be easy to add new bucket types.

Here are the most relevant changes:
- `gcloud.py` will now instantiate and run a PubSub subscriber or GCS bucket based on the value provided by the `--instance_type` argument.
- `integration.py` now contains a generic class from which the pubsub subscriber and the different GCS buckets inherit.
- The `wodles/gcloud/pubsub/` folder has been created. Inside are the classes needed for pubsub. Currently only `subscriber.py` is required.
- The `wodles/gcloud/buckets/` folder has been created. Inside is the generic `WazuhGCloudBucket` class in `bucket.py`, from which all bucket classes should inherit. Currently we only have `GCSAccessLogs`.

Here is how the new hierarchy looks like:
image

### Changes in Core parser

The core parser has been updated to be able to read and process the new configuration needed for `gcp-bucket`. Here is an example configuration:

```
 <gcp-bucket>
    <run_on_start>no</run_on_start>
    <interval>1m</interval>
    <logging>debug</logging>

    <bucket type="access_logs">
        <name>framework-test-bucket</name>
        <credentials_file>wodles/gcp/credentials.json</credentials_file>
        <only_logs_after>2021-JUN-01</only_logs_after>
        <path>access_logs/</path>
        <remove_from_bucket> no </remove_from_bucket>
    </bucket>
</gcp-bucket>
```

The new `gcp-bucket` tag is used, keeping the old `gcp-pubsub` tag for pubsub integration. Changes in the parser for `gcp-bucket` do not affect `gcp-pubsub`.

It is possible to configure more than one bucket in the gcp-bucket configuration:

```
 <gcp-bucket>
    <run_on_start>no</run_on_start>
    <interval>1m</interval>
    <logging>debug</logging>

    <bucket type="access_logs">
        <name>framework-test-bucket</name>
        <credentials_file>wodles/gcp/credentials.json</credentials_file>
        <only_logs_after>2021-JUN-01</only_logs_after>
        <path>access_logs/</path>
        <remove_from_bucket>no</remove_from_bucket>
    </bucket>

    <bucket type="access_logs">
        <name>framework-test-bucket-2</name>
        <credentials_file>wodles/gcp/credentials.json</credentials_file>
        <only_logs_after>2021-JUN-01</only_logs_after>
        <remove_from_bucket>yes</remove_from_bucket>
    </bucket>

</gcp-bucket>
```

## New Features

The new GCS bucket support requires a database. This database is located in `wodles/gcloud/gcloud.db` and it will store the name of the last processed blobs, as well as their creation date.

Due to the limitations of the google cloud storage library, it is not possible to narrow down which blobs (the files stored in a GCS Bucket) to list using the `list_blobs` function. That is why in each iteration `lists_blobs` will return the full list of blobs. To avoid reprocessing logs due to this, creation time and the database are used.

At the end of each execution of the module for a bucket, the creation time of the last blob that was processed, as well as its name, will be stored in the database. In case more than one blob shares the same creation time, the name of all blobs with that creation time will be stored.

This would be a summary of the flow: 
1. The complete list of blobs is downloaded and each blob is taken into account.
2. If the creation time of a blob is older than the creation time value stored in the DB, the file is ignored.
3. If the creation time of a blob is newer than the creation time value stored in the DB, the file will be processed.
If the creation time of a blob is equal to the creation time stored in the DB, a check is made to see if the name of the blob in question is found in the DB. If it is listed, it means that the blob has already been processed, so it shall be ignored. If it does not appear in the list, the blob will be processed.

At the end of the execution, the creation time of the last processed blob or blobs, as well as their names, will be inserted in the DB.

## How to test this

This module can be executed by Wazuh if we apply the corresponding configuration (like the previous example) in `ossec.conf`. It can also be executed manually using the following command:
```
/var/ossec/wodles/gcloud/gcloud.py -t access_logs -b framework-test-bucket -c /var/ossec/wodles/gcloud/credentials.json  -l 1
```
Finally, the pubsub integration can be also tested using the following command:
```
/var/ossec/wodles/gcloud/gcloud -t pubsub -p wazuh-dev-258815 -s test_gcloud_blog -c /var/ossec/wodles/gcloud/credentials.json -l 1
```

### CODE REVIEW

File: wodles/gcloud/tools.py [hunk 1]
```diff
@@ -93,6 +123,63 @@ def get_file_logger(output_file: str, level: int = 3) -> logging.Logger:
...
+            allparts.insert(0, parts[1])
+            break
+        else:
+            abs_path = parts[0]
+            allparts.insert(0, parts[1])
+
+    wazuh_path = ''
+    try:
+        for i in range(0, allparts.index('wodles')):
+            wazuh_path = os.path.join(wazuh_path, allparts[i])
```
Comment (vicferpoy) [inline wodles/gcloud/tools.py:162]:
```suggestion
        wazuh_path = os.path.join(wazuh_path, *[allparts[i] for i in range(allparts.index('wodles'))])
```

An elegant one-liner for that part would be this.

**Review by vicferpoy — CHANGES_REQUESTED** [review]

File: wodles/gcloud/buckets/access_logs.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    def __init__(self, credentials_file: str, logger: logging.Logger, **kwargs):
+        """Class constructor.
+
+        Parameters
+        ----------
+        credentials_file : str
+            Path to credentials file
+        logger : logging.Logger
+            Logger to use
+        """
```
[the remark this answers is not in the document]
Comment (CarlosRS9) [reply wodles/gcloud/buckets/access_logs.py:31]: done

File: wodles/gcloud/buckets/access_logs.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        self.db_table_name = "access_logs"
+        WazuhGCloudBucket.__init__(self, credentials_file, logger, **kwargs)
```
[the remark this answers is not in the document]
Comment (CarlosRS9) [reply wodles/gcloud/buckets/access_logs.py]: Done

File: wodles/gcloud/gcloud.py [hunk 4]
```diff
@@ -9,36 +9,50 @@
 """This module processes events from a Google Cloud subscription."""
 
 import tools
```
[the remark this answers is not in the document]
Comment (CarlosRS9) [reply wodles/gcloud/gcloud.py:11]: This is being used

File: wodles/gcloud/buckets/bucket.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+                                            prefix ='{prefix}'
+                                        ORDER BY
+                                            blob_name;"""
+
+    def check_permissions(self):
+        """Check if the Service Account has access to the bucket."""
+        try:
+            self.bucket = self.client.get_bucket(self.bucket_name)
+        except google_exceptions.NotFound:
+            raise Exception(f'The bucket "{self.bucket_name}" does not exists.')
```
[the remark this answers is not in the document]
Comment (CarlosRS9) [reply wodles/gcloud/buckets/bucket.py]: this is the way the module already handled these exceptions before. We plan to improve this in the future and probably implement an custom exception class for the wodles similar to the one we use in Framework. Unfortunately it is not in the scope of this issue, but we will look into it soon.

**Review by vicferpoy — DISMISSED** [review]
The Python changes LGTM.

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**